# Method 3(a): SVM using Simulated Annealing

In [ ]:
# install Kaggle
! pip install kaggle

In [ ]:
! mkdir ~/.kaggle
!cp kaggle.json ~/.kaggle/

mkdir: cannot create directory ‘/root/.kaggle’: File exists


In [ ]:
! chmod 600 ~/.kaggle/kaggle.json

In [ ]:
# download the dataset from Kaggle
! kaggle datasets download pcbreviglieri/pneumonia-xray-images

 99% 1.12G/1.14G [00:09<00:00, 113MB/s] 
100% 1.14G/1.14G [00:09<00:00, 130MB/s]


In [ ]:
! unzip pneumonia-xray-images
clear_output()

In [ ]:
# import required libraries
import numpy as np
import os
import PIL
import PIL.Image
import tensorflow as tf
import tensorflow_datasets as tfds
from matplotlib import image
from matplotlib import pyplot
from os import listdir
from numpy import save
from tensorflow.keras.utils import load_img
from tensorflow.keras.utils import img_to_array

In [ ]:
# resize opacity images from training set

folder = '/content/train/opacity'
photos_op1 = list()

n1 = 200
n2 = 200

# enumerate files in the directory
for file in listdir(folder):
    # load image
    photo = load_img( '/content/train/opacity/'+ file, target_size=(n1, n2), color_mode='grayscale')
    # convert to numpy array
    photo = img_to_array(photo)
    # store in the list of photos
    photos_op1.append(photo)

# convert list of photos and labels to numpy arrays
photos_op = np.array(photos_op1[:1000])

# label = -1 for opacity cases
labels_op = np.array([-1.0]*1000)

op_train = list()
for i in range(1000):
  op_train.append(photos_op[i].reshape(-1))
op_train = np.array(op_train)
op_train = op_train/255.0

print(op_train.shape)

(1000, 40000)


In [ ]:
# resize normal images from training set
folder = '/content/train/normal'
photos_no = list()

# enumerate files in the directory
for file in listdir(folder):
    # load image
    photo = load_img( '/content/train/normal/'+ file, target_size=(n1, n2), color_mode='grayscale')
    # convert to numpy array
    photo = img_to_array(photo)
    # store in the list of photos
    photos_no.append(photo)

# convert list of photos and labels to numpy arrays
photos_no = np.array(photos_no[:1000])

# label = +1 for normal cases
labels_no = np.array([1.0]*1000)

no_train = list()
for i in range(1000):
  no_train.append(photos_no[i].reshape(-1))
no_train = np.array(no_train)
no_train = no_train/255.0

print(no_train.shape)

(1000, 40000)


In [ ]:
# for normal case images in test data set
folder = '/content/val/normal'
photos_test_no = list()

# enumerate files in the directory
for file in listdir(folder):
    # load image
    photo = load_img( '/content/val/normal/'+ file, target_size=(n1, n2), color_mode='grayscale')
    # convert to numpy array
    photo = img_to_array(photo)
    # store in the list of photos
    photos_test_no.append(photo)

# convert list of photos and labels to numpy arrays
photos_test_no = np.array(photos_test_no)

# label = +1 for normal cases
labels_test_no = np.array([1.0]*photos_test_no.shape[0])

no_test = list()
for i in range(267):
  no_test.append(photos_test_no[i].reshape(-1))
no_test = np.array(no_test)
no_test = no_test/255.0

print(no_test.shape)

(267, 40000)


In [ ]:
# for opacity case test data take extra images from train folder that are not used for training

# convert list of photos and labels to numpy arrays
photos_test_op = np.array(photos_op1[1000:2000])

# label = -1 for normal cases
labels_test_op = np.array([-1.0]*photos_test_op.shape[0])

op_test = list()
for i in range(1000):
  op_test.append(photos_test_op[i].reshape(-1))
op_test = np.array(op_test)
op_test = op_test/255.0

print(op_test.shape)

(1000, 40000)


In [ ]:
# divide all images and labels in sets of 50 each
train = np.empty((40, 50, 40000))
lab = np.empty((40, 50))
for i in range(40):
    train[i] = np.append(op_train[25*i:25*(i+1)], no_train[25*i:25*(i+1)], axis = 0)
    lab[i] = np.append(labels_op[25*i:25*(i+1)], labels_no[25*i:25*(i+1)])

print(train.shape, lab.shape)

(40, 50, 40000) (40, 50)


In [ ]:
# define RBF kernel
import math
def Kernel(mat1, mat2, sigma):
  norm=np.linalg.norm(mat1-mat2)
  k = -(0.5*norm**2)/(sigma**2)
  return np.exp(k)

In [ ]:
# evaluate kernel matrix for all image sets
x=list() # x is the list of all 40 kernel matrices
for k in range(40):
  a = np.empty((50,50))
  for i in range(50):
    for j in range(50):
      a[i][j]=Kernel(train[k][i],train[k][j],50)
  x.append(a)

In [ ]:
# Computing the QUBO matrix for each SVM

I=np.identity(50)
p=[1,2]
pp=np.kron(I,p)

Q=list() # Q is the list of 40 QUBO matrices
for k in range(40):
 y = np.empty((50,50))
 for i in range(50):
   for j in range(50):
     y[i][j] = lab[k][i]*lab[k][j]
 xy=np.multiply(x[k],y)
 q=np.dot(pp.T,np.dot(xy,pp))*0.5
 q=q-np.diag(np.dot(pp.T,np.ones(50)))
 Q.append(q)

In [ ]:
!pip install dwave-ocean-sdk

In [ ]:
# Import the Dwave packages dimod and neal
import dimod
import neal
# Import Matplotlib to generate plots
import matplotlib.pyplot as plt
# Import numpy and scipy for certain numerical calculations below
import numpy as np
from scipy.special import gamma
import math
from collections import Counter
import pandas as pd
from itertools import chain
import time
import networkx as nx

In [ ]:
# perform annealing for 40 SVMs
X=list()
for i in range(40):
 model = dimod.BinaryQuadraticModel.from_qubo(Q[i])
 simAnnSampler = neal.SimulatedAnnealingSampler()
 simAnnSamples = simAnnSampler.sample(model, num_reads=1000)
 energies = simAnnSamples.data_vectors['energy']
 mit = min(energies)
 print(mit)

 # save the values of lagrange multipliers that give minimum energy
 for sample,energy, num_occ in simAnnSamples.data(['sample','energy','num_occurrences']):
  if (energy==mit):
       X.append(np.array(list(sample.values())))
       break

-48.582335896695234
-51.249359517909966
-52.7463342848813
-41.53781051220205
-51.39624957689726
-50.35390545362901
-46.129780567863335
-49.30504350272377
-51.670278269326296
-43.9087314688096
-52.29436522325883
-62.31872046506761
-51.593542276775615
-53.84048409029058
-55.17950219125588
-48.728202246333815
-51.49603306885369
-58.22383714898323
-48.82181459933813
-60.08410459860645
-49.902813999939966
-57.35296895074884
-53.95837855278793
-51.93199226612727
-43.51024787779842
-52.474539708082844
-53.21527981559495
-59.978989105601414
-47.9574655412306
-61.632138234478276
-53.71517758141279
-58.512915119488724
-50.67362139713838
-49.651816688113925
-58.31755706848397
-48.64272272954454
-48.06333820143813
-45.591220025129324
-54.0797065119819
-52.977292941037106


In [ ]:
# convert values from binary to integers
xa=list()
for i in range(40):
  xa.append(np.matmul(pp,X[i].T))

In [ ]:
# calculate bias for each SVM and store values in a list
bias=list()
for i in range(40):
 b=0
 c=0
 for n in range(50):
   a=0
   for m in range(50):
    a=a+xa[i][m]*lab[i][m]*x[i][m][n]
   b=b+xa[i][n]*(3-xa[i][n])*(lab[i][n]-a)
   c=c+xa[i][n]*(3-xa[i][n])
 b = b/c
 bias.append(b)

In [ ]:
# testing the model on opacity images from test data
yy=list()
for i in range(1000):
  a=0.00
  tes=list()
  for j in range(40):
   for k in range(50):
    a=a+xa[j][k]*lab[j][k]*Kernel(op_test[i],train[j][k],50)
   a=a+bias[j]
   if a<0:
    tes.append(-1)
   else:
    tes.append(1)
  yy.append(max(set(tes), key=tes.count))

In [ ]:
count = 0
for i in range(1000):
  if yy[i]==-1:
    count = count + 1
print(count)

927


In [ ]:
# testing the model on normal images from test data
yy=list()
for i in range(267):
  a=0.00
  tes=list()
  for j in range(40):
   for k in range(50):
    a=a+xa[j][k]*lab[j][k]*Kernel(no_test[i],train[j][k],50)
   a=a+bias[j]
   if a<0:
    tes.append(-1)
   else:
    tes.append(1)
  yy.append(max(set(tes), key=tes.count))

In [ ]:
count = 0
for i in range(267):
  if yy[i]==1:
    count = count + 1
print(count)

245


Thus, the model has labelled 927 out of 1000 opacity images and 245 out of 267 normal images correctly. The results can be expressed in a confusion matrix as follows:

|  | Actual +ve | Actual -ve |
| --- | --- | --- |
| **Predicted +ve** | 927 | 22 |
| **Predicted -ve** | 73 | 245 |

Here, positive is opacity and negative is normal.